source/common/http/filter_manager.cc

```diff
@@ -1671,7 +1714,14 @@ Buffer::InstancePtr ActiveStreamEncoderFilter::createBuffer() {
 Buffer::InstancePtr& ActiveStreamEncoderFilter::bufferedData() {
   return parent_.buffered_response_data_;
 }
-bool ActiveStreamEncoderFilter::complete() { return parent_.state_.local_complete_; }
+bool ActiveStreamEncoderFilter::complete() {
+  // This is used for determining end_stream flag when iterating encoder filter chain.
+  // When the upstream half close is enabled the local complete may not be set when
+  // end stream is observed on the encode path in case upstream half closes before the
+  // downstream.
+  return parent_.allow_upstream_half_close_ ? parent_.state_.encoder_end_stream_
+                                            : parent_.state_.local_complete_;
```

Comment:
I think we may need a rename / comment PR before we land this because I think this would be more clear if local_complete were more clear in the first place.

Reply:
or after? (when we consider the urgency of this issue)

Reply:
I have updated comments for local_complete_ and encoder_end_stream_ flags